### PR TITLE
remove perf optimization that cause gigantic memory to be held onto m…

### PR DESCRIPTION
**Pull request by heejaechang**
…emory.

### Customer scenario

Users open a big solution and working on a big file, VS memory usages go up really high for a while and go down and repeat. increasing possibility of OOM.

### Bugs this fixes

https://devdiv.visualstudio.com/DevDiv/_workitems/edit/605655

this is second fix for the same issue (https://github.com/dotnet/roslyn/pull/27718)

### Workarounds, if any

work on small files and type slowly so that roslyn can catch up to latest solution faster.

### Risk

Low. it won't affect functionality or reliability. 

### Performance impact

Low, theoretically, we can create logically same solution twice which might have been cached before and same solution is reused. this is typical trade off of memory for cache vs more computation.

According to RPS, removing this cache doesn't regress any perf scenario. (http://ddweb/dashboard/vsbuild/detail/val/27814.9000.hechang.180614.014123) and we have a dump from customer where it shows this cache holding onto several hundreds meta bytes in memory. 

### Is this a regression from a previous update?

No

### Root cause analysis

basically this Solution/DocumentBranch strongly holding onto next State causes
old roslyn snapshot to make every snapshot after it to stay in memory.

or worse, it causes a lot of forked solution to be left alive in memory until
certain snapshot goes away.

...

Roslyn solution is designed to be not strongly connected to other solution. meaning you can freely create multiple new solution from same solution or compare 2 solutions regardless where it came from.
when you hold onto solution, you expect everything in that snapshot to alive, but not every single snapshot after that to be alive.

I think it is fine that internal implementation to have some information on other solution to make some common functionality faster, but the way it does, strongly hold onto next solution snapshot, seems too much.

### How was the bug found?

UIDelay watson.

### CODE REVIEW

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure how can I run val build from roslyn PR? I would like to get some RPS numbers.

thank you.

**Comment by Therzok** [comment]
Seen this in VSMac also.

The problem seemed to come from holding into a [DocumentSpecificOptionSet](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Options/OptionServiceFactory.cs,162) instance

image

**Comment by heejaechang** [comment]
@Therzok @jasonmalinowski looking at the code, looks like DocumentSpecificOptionSet doesn't need to hold onto Document? _document is passed around, but no one actually uses it.

**Comment by heejaechang** [comment]
@Therzok opened issue.

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs [hunk 1]
```diff
@@ -393,42 +393,10 @@ public new DocumentState UpdateText(SourceText newText, PreservationMode mode)
                 throw new ArgumentNullException(nameof(newText));
             }
 
-            // check to see if this docstate has already been branched before with the same text.
-            // this helps reduce duplicate parsing when typing.
-            if (mode == PreservationMode.PreserveIdentity)
-            {
-                var br = _firstBranch;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: do we still need the _firstBranch field anymore?

**Comment by CyrusNajmabadi** [comment]
I love seeing all the branching code go away.  This code was always nearly impossible to make heads or tails of.  If we do need some sort of optimization here for certain types of patterns, i'd prefer something less invasive, and far clearer in terms of how it works and what contracts it requires.

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: Oh, n/m, i see it was removed below.  bleagh for code that intermixes fields and nested types :D

**Comment by CyrusNajmabadi** [comment]
Can we potentially remove the other bits of branching code?  with all the branches/conditional-tables/etc. it's very hard to make sense of things.  For example:

```c#
private static readonly ConditionalWeakTable<Workspace, ConditionalWeakTable<BranchId, Dictionary<ProjectId, CompilationSet>>> s_map =
    new ConditionalWeakTable<Workspace, ConditionalWeakTable<BranchId, Dictionary<ProjectId, CompilationSet>>>();
```

**Comment by heejaechang** [comment]
@CyrusNajmabadi we have code to use SpeculativeSemanticModel in completionset and semantic tagger. I am +1 on removing those as well.

**Comment by heejaechang** [comment]
by the way, the Branch thing we deleted here and BranchId is completely unrelated thing. they both happen to have term Branch in it. it being used as same meaning as Branch in git.

**Comment by Therzok** [comment]
What is the status on this?

**Comment by heejaechang** [comment]
@Therzok sorry got delayed. will run RPS today.

**Comment by heejaechang** [comment]
retest this please

**Comment by heejaechang** [comment]
ran RPS test - https://devdiv.visualstudio.com/DevDiv/Default/_git/VS/pullrequest/120707?_a=overview - all passed without any regressions.

**Comment by heejaechang** [comment]
can I get some code review? @dotnet/roslyn-ide @sharwell @mavasani @ivanbasov @JieCarolHu @jasonmalinowski 

tagging @jinujoseph to see whether we want this for 15.8

we are talking about several hundreds mega bytes of memory so I think it is worth for 15.8

tagging @Therzok since he was seeing same issue in mac VS.

**Comment by Therzok** [comment]
:+1: from me!

**Comment by heejaechang** [comment]
retest windows_release_vs-integration_prtest please

**Comment by heejaechang** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by heejaechang** [comment]
retest windows_debug_spanish_unit32_prtest  please

**Comment by heejaechang** [comment]
retest this please

**Comment by heejaechang** [comment]
retest windows_release_vs-integration_prtest please

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.Preview5

**Pull request merged**